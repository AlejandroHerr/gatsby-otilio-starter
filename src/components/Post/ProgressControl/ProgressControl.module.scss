@import '../../../colors';

.progress_control_bar {
  pointer-events: none;
}

.progress_control_bar__bar {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 2px;
  background: $primary-color;
  z-index: 1031;
}

.progress_control_bar--finished {
  opacity: .2;
}

.progress_control_bar--hide {
  opacity: 0;
}

.progress_control_bar__peg {
  display: block;
  position: absolute;
  right: 0;
  width: 100px;
  height: 100%;
  box-shadow: 0 0 10px $primary-color, 0 0 5px $primary-color;
  transform: rotate(3deg) translate3d(0, -4px, 0);
  opacity: 1;
}
