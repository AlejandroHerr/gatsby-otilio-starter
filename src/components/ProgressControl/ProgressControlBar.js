import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './ProgressControl.module.scss';

const ProgressControlBar = ({
  finished, progress, speed, mapPropsToStyle,
}) => (
  <div className={classnames(styles.progress_control_bar, {
        [styles['progress_control_bar--finished']]: finished,
        [styles['progress_control_bar--hide']]: progress <= 0,
      })}
  >
    <div
      className={styles.progress_control_bar__bar}
      style={mapPropsToStyle({ finished, progress, speed })}
    >
      {(progress > 0 && (<div className={styles.progress_control_bar__peg} />)) || null}
    </div>
  </div>
);

ProgressControlBar.defaultProps = {
  finished: false,
};

ProgressControlBar.propTypes = {
  finished: PropTypes.bool,
  progress: PropTypes.number.isRequired,
  speed: PropTypes.number.isRequired,
  mapPropsToStyle: PropTypes.func.isRequired,
};

export default ProgressControlBar;
