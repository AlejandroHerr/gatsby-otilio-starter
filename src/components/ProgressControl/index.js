// @flow
import ProgressControl from './ProgressControl';

export default ProgressControl;
