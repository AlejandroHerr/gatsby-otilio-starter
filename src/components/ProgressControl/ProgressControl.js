// @flow
/* global window,document */
import React, { PureComponent, type Node } from 'react';

import ProgressControlBar from './ProgressControlBar';

type PropsType = {
  children: Node,
  speed: number,
};

type StateType = {
  finished: boolean,
  progress: number,
};

export const defaultMapPropsToStyle = ({ progress, speed }: {progress: number, speed: number}) => ({
  width: `${progress}%`,
  transition: `transform ${speed}ms ease`,
});

export default class ProgressControl extends PureComponent<PropsType, StateType> {
  container: ?HTMLDivElement;
  calculateProgress: () => void;

  static defaultProps: {
    speed: number,
  }

  constructor(props: PropsType) {
    super(props);

    this.state = {
      finished: false,
      progress: 0,
    };

    this.container = null;
    this.calculateProgress = this.calculateProgress.bind(this);
  }

  componentDidMount() {
    document.addEventListener('scroll', this.calculateProgress);
    window.addEventListener('resize', this.calculateProgress);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.calculateProgress);
    window.removeEventListener('resize', this.calculateProgress);
  }

  calculateProgress() {
    if (this.container) {
      const { innerHeight, pageYOffset } = window;
      const { clientHeight, offsetTop } = this.container;

      const progress = (100 * pageYOffset) / ((offsetTop + clientHeight) - innerHeight);

      this.setState({ finished: progress > 100, progress: Math.min(progress, 100) });
    }
  }

  render() {
    const { children, speed } = this.props;
    const { finished, progress } = this.state;

    return (
      <div ref={(ref) => { this.container = ref; }}>
        <ProgressControlBar
          finished={finished}
          progress={progress}
          speed={speed}
          mapPropsToStyle={defaultMapPropsToStyle}
        />
        {children}
      </div>
    );
  }
}

ProgressControl.defaultProps = {
  speed: 200,
};
