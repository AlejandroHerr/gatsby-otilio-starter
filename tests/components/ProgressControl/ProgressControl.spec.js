/* globals window,document */
import React from 'react';
import { mount } from 'enzyme';

import ProgressControl, { defaultMapPropsToStyle } from '../../../src/components/ProgressControl/ProgressControl';
import ProgressControlBar from '../../../src/components/ProgressControl/ProgressControlBar';

const TestComponent = () => (<div id="testComponent" />);

const defaultProps = {
  finished: false,
  progress: 0,
  speed: 10,
  mapPropsToStyle: arg => arg,
};

const setup = (props = {}) => {
  const progressControl = mount(
    <ProgressControl {...defaultProps} {...props}>
      <TestComponent />
    </ProgressControl>,
  );

  return {
    progressControl,
    progressControlBar: progressControl.find(ProgressControlBar),
    testComponent: progressControl.find(TestComponent),
  };
};

describe('ProgressControl', () => {
  describe('ProgressControl', () => {
    it('should render ProgressControl', () => {
      const { progressControl, progressControlBar, testComponent } = setup();

      const state = progressControl.state();

      expect(state).toEqual({
        finished: false,
        progress: 0,
      });
      expect(progressControlBar).toHaveLength(1);
      expect(progressControlBar.props()).toEqual({
        finished: state.finished,
        progress: state.progress,
        speed: defaultProps.speed,
        mapPropsToStyle: defaultMapPropsToStyle,
      });
      expect(testComponent).toHaveLength(1);
      expect(progressControl).toMatchSnapshot();
    });
    it('should listen to scroll and resize events', () => {
      document.addEventListener = jest.fn();
      window.addEventListener = jest.fn();
      document.removeEventListener = jest.fn();
      window.removeEventListener = jest.fn();

      const { progressControl } = setup();
      const { calculateProgress } = progressControl.instance();

      expect(document.addEventListener).toHaveBeenCalledWith('scroll', calculateProgress);
      expect(window.addEventListener).toHaveBeenCalledWith('resize', calculateProgress);

      progressControl.unmount();

      expect(document.removeEventListener).toHaveBeenCalledWith('scroll', calculateProgress);
      expect(window.removeEventListener).toHaveBeenCalledWith('resize', calculateProgress);
    });
    it('should calculate the right progress', () => {
      const { progressControl } = setup();
      const { container, calculateProgress } = progressControl.instance();

      calculateProgress();

      expect(progressControl.state()).toEqual({
        finished: false,
        progress: (100 * window.pageYOffset)
          / ((container.offsetTop + container.clientHeight) - window.innerHeight),
      });
    });
  });
  describe('defaultMapPropsToStyle', () => {
    it('should return the right styles', () => {
      expect(defaultMapPropsToStyle({ progress: 20, speed: 40 })).toEqual({
        width: '20%',
        transition: 'transform 40ms ease',
      });
    });
  });
});
