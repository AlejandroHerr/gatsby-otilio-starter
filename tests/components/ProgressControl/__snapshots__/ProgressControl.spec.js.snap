// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressControl ProgressControl should render ProgressControl 1`] = `
<ProgressControl
  finished={false}
  mapPropsToStyle={[Function]}
  progress={0}
  speed={10}
>
  <div>
    <ProgressControlBar
      finished={false}
      mapPropsToStyle={[Function]}
      progress={0}
      speed={10}
    >
      <div
        className="progress_control_bar progress_control_bar--hide"
      >
        <div
          className="progress_control_bar__bar"
          style={
            Object {
              "transition": "transform 10ms ease",
              "width": "0%",
            }
          }
        />
      </div>
    </ProgressControlBar>
    <TestComponent>
      <div
        id="testComponent"
      />
    </TestComponent>
  </div>
</ProgressControl>
`;
